pkg/controllers/user/pipeline/controller/clusterpipeline/clusterpipelinesyncer.go

```diff
@@ -0,0 +1,108 @@
+package clusterpipeline
+
+import (
+	"github.com/pkg/errors"
+	"github.com/rancher/rancher/pkg/controllers/user/pipeline/utils"
+	"github.com/rancher/types/apis/core/v1"
+	"github.com/rancher/types/apis/management.cattle.io/v3"
+	"github.com/rancher/types/config"
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type Syncer struct {
+	cluster    *config.UserContext
+	namespaces v1.NamespaceInterface
+}
+
+func Register(cluster *config.UserContext) {
+	clusterPipelines := cluster.Management.Management.ClusterPipelines("")
+	clusterPipelineSyncer := &Syncer{
+		cluster: cluster,
+	}
+	clusterPipelines.AddHandler("cluster-pipeline-syncer", clusterPipelineSyncer.Sync)
+}
+
+func (s *Syncer) Sync(key string, obj *v3.ClusterPipeline) error {
+	//ensure clusterpipeline singleton in the cluster
+	utils.InitClusterPipeline(s.cluster)
+
+	if obj.Spec.Deploy {
+		if err := s.deploy(); err != nil {
+			return err
+		}
+	} else {
+		if err := s.destroy(); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (s *Syncer) destroy() error {
+	if err := s.cluster.Core.Namespaces("").Delete(utils.PipelineNamespace, &metav1.DeleteOptions{}); err != nil && !apierrors.IsNotFound(err) {
+
+		logrus.Errorf("Error occured while removing ns: %v", err)
+		return err
+
+	}
+
+	return nil
+}
+
+func (s *Syncer) deploy() error {
+	ns := &corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: utils.PipelineNamespace,
+		},
+	}
+	if _, err := s.cluster.Core.Namespaces("").Create(ns); err != nil && !apierrors.IsAlreadyExists(err) {
+		logrus.Errorf("Error occured while create ns: %v", err)
```

Comment:
No need to log in controllers. Just return the error, it will be logged in the framework.